### ISSUE TITLE
[BUG] Multiplayer Issue - Duplicate Static Player

### ISSUE THREAD

**Issue by Acros5542** [AoTTG-2/AoTTG-2#383]
**Describe the bug**
This bug is one in a series of bugs that arose all at once when the photon servers were updated. They did not exist before these updates and still currently exist on both Alpha 0.1 and Alpha 0.2.
When a second player joins a multiplayer lobby, a clone of him is created at the center of the map that is static (unable to be moved, even by hooking him or being pushed by titans). This duplicate will be invisible until the host types /restart at least once, after that, he will continue to be invisible to everyone except the host. This duplicate player also is targetable by the titans’ AI, usually causing most of the titans to gather around the center of the map. The player is hookable, although he won’t move. Sometimes he appears to be falling in place, sometimes he appears standing on the ground. When he’s standing on the ground, he is killable by the titans, with each death causing a null reference error in the developer console, although when he is killed, he instantly respawns in the same spot.
A null reference error is given in the developer console from the perspective of the second player upon joining the lobby which should be in the log file attached below. Even just sitting in the lobby as the second player will slowly fill your developer console with more and more errors.
If you instead spawn as a Titan instead of a Player as the nonMC, a duplicate static titan will instead spawn in the middle of the map that is completely unkillable. There are a lot of issues that occur after this point, but they will be reported as separate bugs.

**To Reproduce**
1. Create a multiplayer lobby
2. Open a second client of AoTTG2,join this lobby, and spawn in as a player
3. Back on the first client, type /restart
4. Head towards the center of the map (easiest to see on Forest)

**Expected behavior**
No static duplicate player or titan should be appearing at all.

**Screenshots**
![Screenshot (353)](https://user-images.githubusercontent.com/75679898/102844222-f0fc5e00-43d8-11eb-9ea9-a6cce2501da0.png)

**Log file**
https://www.dropbox.com/s/codvxw6y3x1ljfr/Multiplayer%20Duplicate.txt?dl=0

**Contributor instructions**
1. Try to recreate the bug via the reported reproduction steps. If the bug can be reproduced, remove the investigation label. If the bug cannot be reproduced, please add a comment, attach "On Hold" label and inform the submitter that the bug cannot be reproduced. If there is no reply after 2 weeks, the issue can be closed.
2. Follow the default development team procedures (Create a new branch, apply your fix, create a pull request, submit new builds for testing team in dev squad)
3. Once testing team has confirmed the issue is resolved and the pull request is accepted, the issue can be closed.

**Comment by Lithrun** [comment]
This sounds something related to #33, where I might have fixed it. If it's fixed in #33, then I will try to get #33 done asap to include this fix. If it's not fixed in #33, then I will create a separate branch to include a fix for this bug

**Comment by Lithrun** [comment]
Issue occurs on #33 as well, so time to create branch 383 and have some fun

**Comment by Lithrun** [comment]
Issue is caused because a non-MC has OnLevelWasLoaded triggered twice. This for some reason, a non-MC has to load a scene twice. That's not something we want, so let's fix that

**Comment by Acros5542** [comment]
**Referring to the Issue-383 build**
___
**Intro**

It is still possible to recreate this issue, but it is a little more complicated and contingent on certain factors. For the sake of explaining this in the most understandable way I can, I’m going to define some variables:

- Player 1 - The host/MasterClient (MC) of the server
- Player 2 - The first person to connect to the server
- Player 3 - The second person to connect to the server
- Duplicate Static Player - the immovable player model that appears in the center of the map (read the original issue for more details)
___
**Explanation**

Player 1 launches the game and hosts a server.
Player 2 launches the game and connects to this server. Upon connecting, Player 2 is given a Null Reference Exception in the Developer Console. The Duplicate Static Player does not appear.
Upon leaving the server and reconnecting, it is very possible Player 2 is given another different error and fails to connect, instead being returned to the main menu. I believe this error is unrelated, but I will explain more about it below. Upon attempting to connect, this error will not be given and Player 2 will be able to connect successfully. Upon connecting, Player 2 will not be given a Null Reference as received before.
Now, from Player 1’s perspective, the Duplicate Static Player has appeared in the center of the map, and it contains all the same properties as it did in the previous version of this issue (targetable by titans, immovable, immediately respawns upon death)
If Player 1 restarts the game using /restart. The Duplicate Static Player will be deleted (it appears to travel from its location at the center of the map to wherever Player 2 spawned).
As in the previous version, the Duplicate Static Player will never be visible to Player 2, and for all future restarts while Player 2 is connected to the server, the Duplicate Static Player will not reappear for either client.
When Player 3 launches the game and connects to the server, he will experience the same series of events as Player 2. The first connect will not spawn the Duplicate Static Player, and Player 3 will be given a Null Reference Error in the Developer Console. Upon leaving and attempting to reconnect, it is likely he will be given an error and will have to retry to connect, but this does not always occur. Upon reconnecting, the Duplicate Static Player will appear for both Player 1 and Player 2 (it appears for everyone except the person who “created” it by connecting).
___
**Additional Information**

The initial connection, the one which causes the Null Reference and does not spawn the Duplicate Static Player, can be bypassed by connecting to any other lobby prior to connecting to Player 1’s server. It only occurs for the 1st lobby you connect to upon launching the game, and if that lobby is a single player one, no error at all will be given. If this initial connection state is bypassed, the Duplicate Static Player will spawn upon the first connection to Player 1’s server.

Each time the Duplicate Static Player is attempted to be moved (example being when Player 1 attempts to hook it), the player who “created” it by connecting to the server will be given a Null Reference Exception in the Developer Console.
Each time the Duplicate Static Player is “killed” by a titan (since it exhibits the same behavior as the original issue, it dies, but immediately respawns in the same location), Player 1 will be given a Null Reference Exception.
___
**The Connection Error**

The error mentioned above that I believe to be unrelated displays this in the Developer Console:
![Screenshot (449)](https://user-images.githubusercontent.com/75679898/105251037-c4a73f00-5b48-11eb-9c16-2896618652c7.png)
Since this error is encountered while recreating the Duplicate Static Player issue in this build, and since there is still a chance it is related, I will leave my explanation of it here, but it is also the same issue I reported in #387, just now with much more information, so I will also update that issue with these details.

Upon first launching the game, your first attempt to connect to a lobby will never give this error. If this first connection to a lobby is a multiplayer lobby, you will instead be given a Null Reference Exception upon connecting successfully to the lobby. If this lobby is a single player lobby, no error will be given. Upon leaving this initial lobby, for your next attempt to connect to a lobby, you will be given the error pictured above. If you were attempting to connect to a multiplayer lobby, you will be given the error and fail to connect, instead being sent back to the main menu. If you were attempting to connect to a single player lobby, you will successfully connect and then be given the error. Upon leaving this second lobby, for your next attempt to connect to a lobby, you will not be given the error and will successfully connect. From then on, this will cycle following this pattern, with one connect giving the error and failing if it was an attempted connection to a multiplayer lobby, and the next connect not giving the error and succeeding. This can be mix-matched in any way, meaning you can connect to a single player lobby and be given the error, then be guaranteed to not receive the error upon leaving the single player session and attempting to connect to a multiplayer session. Or, if there are multiple multiplayer sessions open, you can attempt to connect to one and be given the error, and then successfully connect to the other session without the error. The error will never occur twice in a row.

**Comment by Acros5542** [comment]
On Hold till multiplayer works.
